aurora-solidity-sdk/src/Types.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: CC-BY-1.0
+pragma solidity ^0.8.17;
+
+/// Basic NEAR promise.
+struct PromiseCreateArgs {
+    /// Account id of the target contract to be called.
+    string targetAccountId;
+    /// Method in the contract to be called
+    string method;
+    /// Payload to be passed to the method as input.
+    bytes args;
+    /// Amount of NEAR tokens to attach to the call. This will
+    /// be charged from the caller in wNEAR.
+    uint128 nearBalance;
+    /// Amount of gas to attach to the call.
+    uint64 nearGas;
+}
+
+enum PromiseArgsVariant
+/// Basic NEAR promise
+{
+    Create,
+    /// NEAR promise with a callback attached.
+    Callback,
+    /// Description of arbitrary NEAR promise. Allows applying combinators
+    /// recursively, multiple action types and batched actions.
+    Recursive
```

Comment:
Yes, that's right. But I describe it as recursive because the combinators can be used to create complex call trees (`then` pawns children, `and` joins multiple concurrent operations). The way this is modelled internally is with a self-referential type (like how you might model other graphs/trees).